src/txn/txn.c

```diff
@@ -1431,33 +1431,21 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
         WT_ERR(__txn_fixup_hs_update(session, hs_cursor));
 
 prepare_verify:
-    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED)) {
-        for (; head_upd != NULL; head_upd = head_upd->next) {
-            /*
-             * Assert if we still have an update from the current transaction that hasn't been
-             * resolved or aborted.
-             */
-            WT_ASSERT_ALWAYS(session,
-              head_upd->txnid == WT_TXN_ABORTED || head_upd->prepare_state == WT_PREPARE_RESOLVED ||
-                head_upd->txnid != txn->id,
+    /*
+     * If we are committing a prepared transaction we can check, to some level, that we resolved the
+     * update chain correctly.
+     */
+    if (EXTRA_DIAGNOSTICS_ENABLED(session, WT_DIAGNOSTIC_PREPARED) && commit) {
+        /*
+         * We need to stop walking as soon as we see an aborted update or an update that doesn't
+         * belong to our transaction as it could be freed concurrently by the update obsolete check
+         * logic.
+         */
```

Comment:
How can you ensure the prepared updates cannot be freed? If we have multiple prepared updates and if they all become globally visible, I think the following ones still can be freed by obsolete check.

Reply:
I believe we're protected by the durable_timestamp on the prepared transaction. I don't think the global visibility point can move beyond it as if the stable timestamp moves beyond it we panic. Here: https://github.com/wiredtiger/wiredtiger/blob/7c9d8955e76ef93ff77976f3b3432af50584117a/src/txn/txn.c#L1800 I'm not 100% on this though, and could be convinced I'm not correct.

Reply:
This check is before the commit. I think after that it is possible the global timestamp can move beyond that but the user should not do that.

Reply:
You mean the check that I am modifying? Yes it's before the commit. Each prepared update will have its durable timestamp set as part of resolution in `__txn_resolve_prepared_update`. In order for those to become globally visible the oldest timestamp must move beyond them. For that to happen concurrently the application needs to move the stable and oldest timestamp, which would cause the panic to fire that I indicated above.

So my reasoning is this:
If they were to become globally visible we would panic, thus they should never become globally visible. 

Do you agree with my reasoning? I am aware that if we were to panic then in theory the update chain could be freed concurrently and we'd hit a heap corruption prior to panicking. I suspect 9 times out of 10 you'd panic before you got the heap corruption as it requires additional updates to be put on the chain as well as the global visibility state moving.

Reply:
I mean the prepared operations are not globally visible when you are in this check

```
if (prepare && txn->durable_timestamp <= txn_global->stable_timestamp)
```

when you reach the verify code then the user moves the oldest timestamp to make the operation globally visible. I think in that case it is possible to hit the heap corruption.

```
I suspect 9 times out of 10 you'd panic before you got the heap corruption as it requires additional updates to be put on the chain as well as the global visibility state moving.
```

I agree that it is very rare to hit heap corruption but it is possible with the above sequence. I still approved the PR because user should not move the global timestamp in this case. It is only a problem in theory.

Reply:
I made a mistake that I thought the check:

```
if (prepare && txn->durable_timestamp <= txn_global->stable_timestamp)
```

is at the start of the commit. Actually it is at the end of the commit. Therefore, it is all good.